// @flatfile/api-beta/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlatfileClient = void 0;
const url_join_1 = __importDefault(require("url-join"));
const environments = __importStar(require("./environments"));
const core = __importStar(require("./core"));
const errors = __importStar(require("./errors"));
const Client_1 = require("./api/resources/spaces/client/Client");
const Client_2 = require("./api/resources/agents/client/Client");
const Client_3 = require("./api/resources/cells/client/Client");
const Client_4 = require("./api/resources/documents/client/Client");
const Client_5 = require("./api/resources/environments/client/Client");
const Client_6 = require("./api/resources/events/client/Client");
const Client_7 = require("./api/resources/files/client/Client");
const Client_8 = require("./api/resources/guests/client/Client");
const Client_9 = require("./api/resources/jobs/client/Client");
const Client_10 = require("./api/resources/records/client/Client");
const Client_11 = require("./api/resources/sheets/client/Client");
const Client_12 = require("./api/resources/users/client/Client");
const Client_13 = require("./api/resources/versions/client/Client");
const Client_14 = require("./api/resources/workbooks/client/Client");
class FlatfileClient {
    token;
    options;
    constructor(options) {
        this.options = {
            environment: options.environment,
            token: async () => {
                if (this.token == null) {
                    const response = await core.fetcher({
                        url: (0, url_join_1.default)(this.options.environment ?? environments.FlatfileEnvironment.Production, "/auth/access-token"),
                        method: "POST",
                        body: {
                            clientId: options.clientId,
                            secret: options.clientSecret,
                        },
                    });
                    if (response.ok) {
                        this.token = response.body?.data?.accessToken;
                    }
                }
                if (this.token == null) {
                    throw new errors.FlatfileError({ message: "Failed to authenticate" });
                }
                return this.token;
            },
        };
    }
    _spaces;
    get spaces() {
        return (this._spaces ??= new Client_1.Spaces(this.options));
    }
    _agents;
    get agents() {
        return (this._agents ??= new Client_2.Agents(this.options));
    }
    _cells;
    get cells() {
        return (this._cells ??= new Client_3.Cells(this.options));
    }
    _documents;
    get documents() {
        return (this._documents ??= new Client_4.Documents(this.options));
    }
    _environments;
    get environments() {
        return (this._environments ??= new Client_5.Environments(this.options));
    }
    _events;
    get events() {
        return (this._events ??= new Client_6.Events(this.options));
    }
    _files;
    get files() {
        return (this._files ??= new Client_7.Files(this.options));
    }
    _guests;
    get guests() {
        return (this._guests ??= new Client_8.Guests(this.options));
    }
    _jobs;
    get jobs() {
        return (this._jobs ??= new Client_9.Jobs(this.options));
    }
    _records;
    get records() {
        return (this._records ??= new Client_10.Records(this.options));
    }
    _sheets;
    get sheets() {
        return (this._sheets ??= new Client_11.Sheets(this.options));
    }
    _users;
    get users() {
        return (this._users ??= new Client_12.Users(this.options));
    }
    _versions;
    get versions() {
        return (this._versions ??= new Client_13.Versions(this.options));
    }
    _workbooks;
    get workbooks() {
        return (this._workbooks ??= new Client_14.Workbooks(this.options));
    }
}
exports.FlatfileClient = FlatfileClient;
