// @flatfile/api-beta/serialization/resources/events/types/Context.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Context = void 0;
const core = __importStar(require("../../../../core"));
exports.Context = core.serialization.object({
    actionName: core.serialization.lazy(async () => (await Promise.resolve().then(() => __importStar(require("../../..")))).ActionName).optional(),
    accountId: core.serialization.lazy(async () => (await Promise.resolve().then(() => __importStar(require("../../..")))).AccountId),
    environmentId: core.serialization.lazy(async () => (await Promise.resolve().then(() => __importStar(require("../../..")))).EnvironmentId),
    spaceId: core.serialization.lazy(async () => (await Promise.resolve().then(() => __importStar(require("../../..")))).SpaceId).optional(),
    workbookId: core.serialization.lazy(async () => (await Promise.resolve().then(() => __importStar(require("../../..")))).WorkbookId).optional(),
    sheetId: core.serialization.lazy(async () => (await Promise.resolve().then(() => __importStar(require("../../..")))).SheetId).optional(),
    sheetSlug: core.serialization.lazy(async () => (await Promise.resolve().then(() => __importStar(require("../../..")))).SheetSlug).optional(),
    versionId: core.serialization.lazy(async () => (await Promise.resolve().then(() => __importStar(require("../../..")))).VersionId).optional(),
    jobId: core.serialization.lazy(async () => (await Promise.resolve().then(() => __importStar(require("../../..")))).JobId).optional(),
    fileId: core.serialization.lazy(async () => (await Promise.resolve().then(() => __importStar(require("../../..")))).FileId).optional(),
    precedingEventId: core.serialization.lazy(async () => (await Promise.resolve().then(() => __importStar(require("../../..")))).EventId).optional(),
});
