// @flatfile/api-beta/dist/api/resources/events/types/EventTopic.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventTopic = void 0;
exports.EventTopic = {
    SpaceAdded: "space:added",
    SpaceRemoved: "space:removed",
    WorkbookAdded: "workbook:added",
    WorkbookRemoved: "workbook:removed",
    UserAdded: "user:added",
    UserRemoved: "user:removed",
    UserOnline: "user:online",
    UserOffline: "user:offline",
    UploadStarted: "upload:started",
    UploadFailed: "upload:failed",
    UploadCompleted: "upload:completed",
    JobStarted: "job:started",
    JobWaiting: "job:waiting",
    JobUpdated: "job:updated",
    JobFailed: "job:failed",
    JobCompleted: "job:completed",
    JobDeleted: "job:deleted",
    RecordsCreated: "records:created",
    RecordsUpdated: "records:updated",
    RecordsDeleted: "records:deleted",
    SheetValidated: "sheet:validated",
    ActionTriggered: "action:triggered",
    FileDeleted: "file:deleted",
};
